boring/src/x509/crl.rs

```diff
@@ -0,0 +1,309 @@
+        X509CRL,
+        ffi::d2i_X509_CRL,
+        ::libc::c_long
+    }
+}
+
+impl X509CRLRef {
+    /// Returns the CRL's last update time
+    ///
+    /// This corresponds to [`X509_CRL_get0_lastUpdate`]
+    ///
+    /// [`X509_CRL_get0_lastUpdate`]: https://www.openssl.org/docs/man1.1.1/man3/X509_CRL_get0_lastUpdate
+    pub fn last_update(&self) -> &Asn1TimeRef {
+        unsafe {
+            let date = ffi::X509_CRL_get0_lastUpdate(self.as_ptr());
+            assert!(!date.is_null());
+            Asn1TimeRef::from_ptr(date as *mut _)
+        }
+    }
+
+    /// Returns the CRL's next update time
+    ///
+    /// This corresponds to [`X509_CRL_get0_nextUpdate`]
+    ///
+    /// [`X509_CRL_get0_nextUpdate`]: https://www.openssl.org/docs/man1.1.1/man3/X509_CRL_get0_nextUpdate
+    pub fn next_update(&self) -> &Asn1TimeRef {
+        unsafe {
+            let date = ffi::X509_CRL_get0_nextUpdate(self.as_ptr());
+            assert!(!date.is_null());
+            Asn1TimeRef::from_ptr(date as *mut _)
+        }
+    }
+
+    /// Returns the CRL's issuer name
+    ///
+    /// This corresponds to [`X509_CRL_get_issuer`]
+    ///
+    /// [`X509_CRL_get_issuer`]: https://www.openssl.org/docs/man1.1.1/man3/X509_CRL_get_issuer
+    pub fn issuer(&self) -> &X509NameRef {
+        unsafe {
+            #[cfg(feature = "fips")]
+            let name = {
+                let crl_info = (*self.as_ptr()).crl;
+                assert!(!crl_info.is_null());
+                *crl_info.issuer
+            };
+
+            #[cfg(not(feature = "fips"))]
+            let name = ffi::X509_CRL_get_issuer(self.as_ptr());
+
+            assert!(!name.is_null());
+            X509NameRef::from_ptr(name)
+        }
+    }
+
+    /// Returns the CRL's extensions
+    ///
+    /// This corresponds to [`X509_CRL_get0_extensions`]
+    ///
+    /// [`X509_CRL_get0_extensions`]: https://www.openssl.org/docs/man1.1.1/man3/X509_CRL_get0_extensions
+    pub fn extensions(&self) -> Option<&StackRef<X509Extension>> {
+        unsafe {
+            #[cfg(not(feature = "fips"))]
+            let extensions = ffi::X509_CRL_get0_extensions(self.as_ptr());
+            #[cfg(feature = "fips")]
+            let extensions = {
+                let crl_info = (*self.as_ptr()).crl;
+                assert!(!crl_info.is_null());
+                *crl_info.extensions;
+            };
+            if extensions.is_null() {
+                None
+            } else {
+                Some(StackRef::from_ptr(extensions as *mut _))
+            }
+        }
+    }
+
+    /// Returns the revoked certificates in this CRL
+    ///
+    /// This corresponds to [`X509_CRL_get_REVOKED`]
+    ///
+    /// [`X509_CRL_get_REVOKED`]: https://www.openssl.org/docs/man1.1.1/man3/X509_CRL_get_REVOKED
+    pub fn revoked(&self) -> Option<&StackRef<X509Revoked>> {
+        unsafe {
+            #[cfg(not(feature = "fips"))]
+            let revoked = ffi::X509_CRL_get_REVOKED(self.as_ptr());
```

Comment:
In the FIPS boring-sys version these were c macros, but the the struct definitions were in a header that was bind-gened. In the non-fips version, they are functions and the struct definitions are in `internal.h`.

Reply:
Does other code in boring have a similar pattern?

Reply:
Not really! In most places it just conditionally omits the function in question. It seemed not supporting `revoked` would be pretty hindering compared to the existing cases. However, I can see the argument of starting with the more conservative approach.

Reply:
Looks like no. :-/ I guess we could ask upstream whether they want it working with the FIPS variant, or whether it'd be okay to only support the newer API.